Check slice bounds in tsm1 to avoid panics

mark-rushakoff opened the pull request:
I used go-fuzz (WIP branch at https://github.com/mark-rushakoff/influxdb/tree/mr-add-gofuzz) and found a handful of panics.

These all seem to be pretty low level in the TSM1 engine, e.g.:

```
panic: runtime error: slice bounds out of range

goroutine 1 [running]:
panic(0x478920, 0xc82000a150)
	/var/folders/ct/bl4_z3g51ks8239_r2k07v_40000gn/T/go-fuzz-build267795264/src/runtime/panic.go:481 +0x3e6
github.com/influxdata/influxdb/tsdb/engine/tsm1.unpackBlock(0xa0e00a, 0x3b, 0x6c, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/var/folders/ct/bl4_z3g51ks8239_r2k07v_40000gn/T/go-fuzz-build267795264/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/encoding.go:591 +0x121
github.com/influxdata/influxdb/tsdb/engine/tsm1.DecodeStringBlock(0xa0e00a, 0x3b, 0x6c, 0xc8200a0000, 0xc82003ba50, 0xc82003ba00, 0x0, 0x0, 0x0, 0x0, ...)
	/var/folders/ct/bl4_z3g51ks8239_r2k07v_40000gn/T/go-fuzz-build267795264/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/encoding.go:533 +0x332
github.com/influxdata/influxdb/tsdb/engine/tsm1.DecodeBlock(0xa0e009, 0x3c, 0x6d, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/var/folders/ct/bl4_z3g51ks8239_r2k07v_40000gn/T/go-fuzz-build267795264/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/encoding.go:176 +0x1499
github.com/influxdata/influxdb/tsdb/engine/tsm1.(*mmapAccessor).readAll(0xc82000e780, 0xc82000b0d8, 0x8, 0x0, 0x0, 0x0, 0x0, 0x0)
	/var/folders/ct/bl4_z3g51ks8239_r2k07v_40000gn/T/go-fuzz-build267795264/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/reader.go:1020 +0x723
github.com/influxdata/influxdb/tsdb/engine/tsm1.(*TSMReader).ReadAll(0xc820014730, 0xc82000b0d8, 0x8, 0x0, 0x0, 0x0, 0x0, 0x0)
	/var/folders/ct/bl4_z3g51ks8239_r2k07v_40000gn/T/go-fuzz-build267795264/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/reader.go:286 +0x114
github.com/influxdata/influxdb/tsdb/engine/tsm1.FuzzTSMReader(0x80e000, 0x76, 0x200000, 0x0)
	/var/folders/ct/bl4_z3g51ks8239_r2k07v_40000gn/T/go-fuzz-build267795264/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/fuzz.go:103 +0x72a
go-fuzz-dep.Main(0x574760)
	/var/folders/ct/bl4_z3g51ks8239_r2k07v_40000gn/T/go-fuzz-build267795264/src/go-fuzz-dep/main.go:49 +0x14c
main.main()
	/var/folders/ct/bl4_z3g51ks8239_r2k07v_40000gn/T/go-fuzz-build267795264/src/go.fuzz.main/main.go:10 +0x23
```

So the test case I've added may not be exactly right - I'm probably expecting no errors in places where an error is appropriate - but nonetheless the added tests in this PR reproduce several panics. It may be most appropriate to fix the panics separately and close this PR.

/cc @jwilder @benbjohnson

mark-rushakoff commented:
There seems to be some shared state in CacheLoader that results in one of the added tests passing when run on its own but failing (OOMing?) when run as part of the full suite:

```
$ go test -v  -run=Crash ./tsdb/engine/tsm1/
=== RUN   TestCacheLoader_FuzzCrashes
--- PASS: TestCacheLoader_FuzzCrashes (3.79s)
=== RUN   TestTSMReader_FuzzCrashes
--- PASS: TestTSMReader_FuzzCrashes (0.13s)
PASS
ok  	github.com/influxdata/influxdb/tsdb/engine/tsm1	4.090s
```

```
$ go test -v  -run=CacheLoader ./tsdb/engine/tsm1/
=== RUN   TestCacheLoader_LoadSingle
[cacheloader] 2016/07/15 10:56:00 reading file /var/folders/ct/bl4_z3g51ks8239_r2k07v_40000gn/T/tsm1-test032412442/tsm1test997787057, size 53
[cacheloader] 2016/07/15 10:56:00 reading file /var/folders/ct/bl4_z3g51ks8239_r2k07v_40000gn/T/tsm1-test032412442/tsm1test997787057, size 58
[cacheloader] 2016/07/15 10:56:00 file /var/folders/ct/bl4_z3g51ks8239_r2k07v_40000gn/T/tsm1-test032412442/tsm1test997787057 corrupt at position 53, truncating
--- PASS: TestCacheLoader_LoadSingle (0.00s)
=== RUN   TestCacheLoader_LoadDouble
[cacheloader] 2016/07/15 10:56:00 reading file /var/folders/ct/bl4_z3g51ks8239_r2k07v_40000gn/T/tsm1-test970931036/tsm1test472192779, size 89
[cacheloader] 2016/07/15 10:56:00 file /var/folders/ct/bl4_z3g51ks8239_r2k07v_40000gn/T/tsm1-test970931036/tsm1test472192779 corrupt at position 84, truncating
[cacheloader] 2016/07/15 10:56:00 reading file /var/folders/ct/bl4_z3g51ks8239_r2k07v_40000gn/T/tsm1-test970931036/tsm1test518701806, size 0
--- PASS: TestCacheLoader_LoadDouble (0.03s)
=== RUN   TestCacheLoader_LoadDeleted
[cacheloader] 2016/07/15 10:56:00 reading file /var/folders/ct/bl4_z3g51ks8239_r2k07v_40000gn/T/tsm1-test631825269/tsm1test342384720, size 65
[cacheloader] 2016/07/15 10:56:00 reading file /var/folders/ct/bl4_z3g51ks8239_r2k07v_40000gn/T/tsm1-test631825269/tsm1test342384720, size 65
--- PASS: TestCacheLoader_LoadDeleted (0.00s)
=== RUN   TestCacheLoader_FuzzCrashes
signal: killed
FAIL	github.com/influxdata/influxdb/tsdb/engine/tsm1	38.908s
```

I'm going to dig in some more and see what I can find.

mark-rushakoff commented:
The reason for the red build before was that the CacheLoader was OOMing when trying to allocate and populate a slice of over 4 billion references. I removed that test, build is green now.

jwilder commented on tsdb/engine/tsm1/bool.go:
```diff
@@ -97,19 +100,32 @@ type BooleanDecoder struct {
 // This must be called before calling any other methods.
 func (e *BooleanDecoder) SetBytes(b []byte) {
+	if len(b) == 0 {
+		e.err = fmt.Errorf("BooleanDecoder: initialized with empty data")
```
This shouldn't be an error.  Can just return.

jwilder commented on tsdb/engine/tsm1/bool.go:
```diff
@@ -97,19 +100,32 @@ type BooleanDecoder struct {
+
 	e.i++
-	return e.i < e.n
+	return e.i/8 < len(e.b) && e.i < e.n
```
The `e.i/8 < len(e.b)` shouldn't be necessary.  It also adds a division in a hot section of decoder.

jwilder commented on tsdb/engine/tsm1/encoding.go:
```diff
@@ -122,7 +122,10 @@ func BlockCount(block []byte) int {
-	tb, _ := unpackBlock(block[1:])
+	tb, _, err := unpackBlock(block[1:])
+	if err != nil {
+		panic(fmt.Sprintf("BlockCount: error unpacking block: %s", err.Error()))
```
Unrelated to this PR but `BlockCount` should be returning an error, not panicing.

jwilder commented on tsdb/engine/tsm1/float.go:
```diff
@@ -142,6 +142,10 @@ type FloatDecoder struct {
 // SetBytes initializes the decoder with b. Must call before calling Next().
 func (it *FloatDecoder) SetBytes(b []byte) error {
+	if len(b) == 0 {
+		return fmt.Errorf("FloatDecoder: SetBytes called with empty data")
```
Should not return an error here, just reset everything back to initial state.

jwilder commented on tsdb/engine/tsm1/reader.go:
```diff
@@ -768,6 +773,10 @@ func (d *indirectIndex) MarshalBinary() ([]byte, error) {
 // representation of an index.
 func (d *indirectIndex) UnmarshalBinary(b []byte) error {
+	if len(b) == 0 {
+		return fmt.Errorf("indirectIndex: cannot unmarshal from empty slice")
```
Should not be an error, just return.

jwilder commented on tsdb/engine/tsm1/string.go:
```diff
@@ -57,6 +57,10 @@ type StringDecoder struct {
 // This must be called before calling any other method.
 func (e *StringDecoder) SetBytes(b []byte) error {
+	if len(b) == 0 {
+		return fmt.Errorf("StringDecoder: initialized with empty data")
```
Not an error, just return.

mark-rushakoff commented:
@jwilder: removed errors about initializing decoders with an empty slice, removed division in hot path of boolean decoder (and added benchmark + speedup for it too).

Still left the panic in `tsm1.BlockCount` since there was already a panic and it would have been a larger API change.

jwilder commented:
Looks great @mark-rushakoff! Thanks! 👍
